auth/auth_test.go

```diff
@@ -0,0 +1,206 @@
+
+	openidConfig = map[string]any{
+		"issuer":                                s.URL,
+		"authorization_endpoint":                s.URL + "/oauth/authorize",
+		"token_endpoint":                        s.URL + "/oauth/token",
+		"userinfo_endpoint":                     s.URL + "/oauth/userinfo",
+		"jwks_uri":                              s.URL + "/oauth/discovery/keys",
+		"id_token_signing_alg_values_supported": []string{"RS256"},
+	}
+
+	token := jwt.NewWithClaims(
+		jwt.SigningMethodRS256,
+		jwt.MapClaims{
+			"iss":   s.URL,
+			"sub":   subject,
+			"aud":   clientID,
+			"exp":   time.Now().Add(time.Hour).Unix(),
+			"iat":   time.Now().Unix(),
+			"nonce": nonce,
+		},
+	)
+	rawIDToken, err = token.SignedString(rs256)
+	require.NoError(t, err)
+	return s
+}
+
+func TestHandler(t *testing.T) {
+	const (
+		testClientID    = "test-client-id"
+		testCode        = "test-code"
+		testAccessToken = "test-access-token"
+		testSubject     = "018d21f2-0b8d-756a-84f0-13b942a2bae5"
+		testName        = "John Doe"
+		testEmail       = "john.doe@example.com"
+	)
+	someday, err := time.Parse(time.DateOnly, "2021-01-01")
+	require.NoError(t, err)
+	userinfo, err := json.Marshal(
+		map[string]any{
+			"sub":            testSubject,
+			"name":           testName,
+			"email":          testEmail,
+			"email_verified": true,
+			"picture":        "https://example.com/avatar.jpg",
+			"created_at":     someday,
+		},
+	)
+	require.NoError(t, err)
+
+	// Set up the mock service server.
+	mockState := ""
+	mux := http.NewServeMux()
+	mockServiceServer := httptest.NewServer(mux)
+	t.Cleanup(func() { mockServiceServer.Close() })
+	redirectURI := mockServiceServer.URL + "/callback"
+
+	// Set up the mock SAMS server.
+	mockSAMSServer := newMockServer(t, redirectURI, testClientID, testSubject, testCode, testAccessToken, userinfo)
+
+	// Set up auth handlers for the mock service server.
+	h, err := NewHandler(
+		Config{
+			Issuer:         mockSAMSServer.URL,
+			ClientID:       testClientID,
+			ClientSecret:   "test-client-secret",
+			RequestScopes:  []scopes.Scope{scopes.OpenID, scopes.Profile, scopes.Email},
+			RedirectURI:    redirectURI,
+			FailureHandler: DefaultFailureHandler,
+			StateSetter: func(_ http.ResponseWriter, _ *http.Request, state string) error {
+				mockState = state
+				return nil
+			},
+			StateGetter: func(*http.Request) (string, error) {
+				return mockState, nil
+			},
+			StateDeleter: func(http.ResponseWriter, *http.Request) {
+				mockState = ""
+			},
+		},
+	)
+	require.NoError(t, err)
+
+	mux.Handle("/login", h.LoginHandler())
+	mux.Handle("/callback", h.CallbackHandler(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			userInfo := UserInfoFromContext(r.Context())
+			assert.NotNil(t, userInfo.Token)
+			assert.NotNil(t, userInfo.IDToken)
+			err := json.NewEncoder(w).Encode(userInfo)
+			require.NoError(t, err)
+		}),
+	))
+
+	// Simulate authentication flow.
+	client := &http.Client{}
```

Comment:
Nit: Why don't we tease this out into a separate `t.Run("SimulateAuthFlow", func(t *testing.T) {` block.

And then add another unit test to sanity check some of the error modes. e.g.

```go
StateSetter = func(...) { return nil, errors.New("setter error")
...
Confirm that the ErrorHandler was invoked with the expected error
```

I understand that it might be more trouble than we'd want to deal with, so feel free to push back if you don't think that would be valuable enough to add support for.

Reply:
Added a case for invoking FailureHandler in https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/8/commits/633f38688fe192c3a381aba40ecb3592219cf6fe